x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportCreateEnrollmentTokenAction.java

```diff
@@ -0,0 +1,144 @@
+
+    public static final long ENROLL_API_KEY_EXPIRATION_SEC = 30*60;
+
+    private final ApiKeyGenerator generator;
+    private final SecurityContext securityContext;
+    private final Environment environment;
+    private final NodeService nodeService;
+    private final SSLService sslService;
+
+    @Inject
+    public TransportCreateEnrollmentTokenAction(TransportService transportService, ActionFilters actionFilters, ApiKeyService apiKeyService,
+                                                SecurityContext context, CompositeRolesStore rolesStore,
+                                                NamedXContentRegistry xContentRegistry, Environment environment, NodeService nodeService,
+                                                SSLService sslService) {
+        super(CreateEnrollmentTokenAction.NAME, transportService, actionFilters, CreateEnrollmentTokenRequest::new);
+        this.generator = new ApiKeyGenerator(apiKeyService, rolesStore, xContentRegistry);
+        this.securityContext = context;
+        this.environment = environment;
+        this.nodeService = nodeService;
+        this.sslService = sslService;
+    }
+
+    @Override
+    protected void doExecute(Task task, CreateEnrollmentTokenRequest request,
+                             ActionListener<CreateEnrollmentTokenResponse> listener) {
+        try {
+            final KeyConfig keyConfig = sslService.getHttpTransportSSLConfiguration().keyConfig();
+            if (keyConfig instanceof StoreKeyConfig == false) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration is not configured with a " +
+                        "keystore"));
+                return;
+            }
+            final List<X509Certificate> caCertificates = ((StoreKeyConfig) keyConfig).x509Certificates(environment).stream()
+                .filter(x509Certificate -> x509Certificate.getBasicConstraints() != -1).collect(Collectors.toList());
+            if (caCertificates.size() != 1) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration Keystore doesn't contain a " +
+                    "single PrivateKey entry where the associated certificate is a CA certificate"));
+            } else {
+                final TimeValue expiration = TimeValue.timeValueSeconds(ENROLL_API_KEY_EXPIRATION_SEC);
+                final List<RoleDescriptor> roleDescriptors = new ArrayList<>(1);
+                final String[] clusterPrivileges = { "enroll" };
+                final RoleDescriptor roleDescriptor = new RoleDescriptor("create_enrollment_token", clusterPrivileges, null, null);
+                roleDescriptors.add(roleDescriptor);
+                CreateApiKeyRequest apiRequest = new CreateApiKeyRequest(UUIDs.base64UUID(), roleDescriptors, expiration);
```

Comment:
Is it intentional to use an UUID as the API key's name? The first argument for `new CreateApiKeyRequest(...)` is the API key name, not its ID. Should we use a more meainingful name? It could at least help with auditing which currently logs only API key name, not ID.

Reply:
I should have been more specific about the auditing: what I meant is when auditing the creation of an API key, only its name is logged, not its ID. Any subsequent requests issued with the API key will have both the API key ID and name audited.

Reply:
I actually had the prefix for the name before and removed it at some point by I don't remember what reasons :). Putting it back.